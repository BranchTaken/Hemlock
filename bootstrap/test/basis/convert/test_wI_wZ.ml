open! Basis.Rudiments
open! Basis

let test () =
  File.Fmt.stdout
  |> (fun formatter ->
    List.fold I128.([of_string "0x8000_0000_0000_0000_0000_0000_0000_0000i128"; of_i64 (-1L);
      of_i64 0L; of_i64 1L; of_string "0x7fff_ffff_ffff_ffff_ffff_ffff_ffff_ffffi128"])
      ~init:formatter ~f:(fun formatter i ->
      formatter
      |> Fmt.fmt "extend_to_zint "
      |> I128.fmt ~alt:true ~base:Fmt.Hex ~pretty:true i
      |> Fmt.fmt " -> "
      |> Zint.fmt ~alt:true ~base:Fmt.Hex ~pretty:true (I128.extend_to_zint i)
      |> Fmt.fmt "\n"
    )
  )
  |> Fmt.fmt "\n"
  |> (fun formatter ->
    List.fold Zint.([of_i64 (-1L); of_i64 0L; of_i64 1L;
      of_string "0x7fff_ffff_ffff_ffff_ffff_ffff_ffff_ffffz";
      of_string "0x8000_0000_0000_0000_0000_0000_0000_0000z";
      of_string "0xffff_ffff_ffff_ffff_ffff_ffff_ffff_ffffz";
      of_string "0x1_0000_0000_0000_0000_0000_0000_0000_0000z";
      of_string "0x1_ffff_ffff_ffff_ffff_ffff_ffff_ffff_ffffz"]) ~init:formatter
      ~f:(fun formatter i ->
        formatter
        |> Fmt.fmt "trunc_of_zint/narrow_of_zint_opt "
        |> Zint.fmt ~alt:true ~base:Fmt.Hex ~pretty:true i
        |> Fmt.fmt " -> "
        |> I128.fmt ~alt:true ~base:Fmt.Hex ~pretty:true (I128.trunc_of_zint i)
        |> Fmt.fmt "/"
        |> (Option.fmt I128.pp) (I128.narrow_of_zint_opt i)
        |> Fmt.fmt "\n"
      )
  )
  |> ignore

let _ = test ()
