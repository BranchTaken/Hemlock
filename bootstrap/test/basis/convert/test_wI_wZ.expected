extend_to_zint -0x8000_0000_0000_0000_0000_0000_0000_0000i128 -> -0x8000_0000_0000_0000_0000_0000_0000_0000z
extend_to_zint -0x1i128 -> -0x1z
extend_to_zint 0x0i128 -> 0x0z
extend_to_zint 0x1i128 -> 0x1z
extend_to_zint 0x7fff_ffff_ffff_ffff_ffff_ffff_ffff_ffffi128 -> 0x7fff_ffff_ffff_ffff_ffff_ffff_ffff_ffffz

trunc_of_zint/narrow_of_zint_opt -0x1z -> -0x1i128/Some -1i128
trunc_of_zint/narrow_of_zint_opt 0x0z -> 0x0i128/Some 0i128
trunc_of_zint/narrow_of_zint_opt 0x1z -> 0x1i128/Some 1i128
trunc_of_zint/narrow_of_zint_opt 0x7fff_ffff_ffff_ffff_ffff_ffff_ffff_ffffz -> 0x7fff_ffff_ffff_ffff_ffff_ffff_ffff_ffffi128/Some 170_141_183_460_469_231_731_687_303_715_884_105_727i128
trunc_of_zint/narrow_of_zint_opt 0x8000_0000_0000_0000_0000_0000_0000_0000z -> -0x8000_0000_0000_0000_0000_0000_0000_0000i128/None
trunc_of_zint/narrow_of_zint_opt 0xffff_ffff_ffff_ffff_ffff_ffff_ffff_ffffz -> -0x1i128/None
trunc_of_zint/narrow_of_zint_opt 0x1_0000_0000_0000_0000_0000_0000_0000_0000z -> 0x0i128/None
trunc_of_zint/narrow_of_zint_opt 0x1_ffff_ffff_ffff_ffff_ffff_ffff_ffff_ffffz -> -0x1i128/None
